---
title: "01 Pre-class setup"
format:
  html:
    code-fold: true
    toc: true
    toc-location: left
editor: visual
---

In this chapter you will learn how to setup your environment either on the remote server or locally.

**Learning objectives**

1. Create directory
2. Copy repo
3. Setup conda environment
4. Download data

# Setup directory

## Login to AUTH cluster (Remote coding only)

Before setting up, there are some necessary steps specific for remote coding which are specific to Windows users. If you are a Linux os MacOS user skip to [section](#unix-linuxmacos-users)

### Install VSCode or MobaXTerm (Windows Users Only)

In order to use SSH (remote host access) to the AUTH computer cluster you will either need to have Windows Subprocesses for Linux [(WSL)](https://learn.microsoft.com/en-us/windows/wsl/install) installed and enabled or use and IDE such as [VSCode](https://code.visualstudio.com/) or [MobaXTerm](https://mobaxterm.mobatek.net/)

### Connect to AUTH cluster

If you are not logged in an AUTH network (e.g. working from home), make sure you have eduVPN enabled. More info [here](https://it.auth.gr/manuals/eduvpn/)

Then open a terminal window or your IDE and type the following:

```bash
ssh [username]@aristotle.it.auth.gr
```

## Download repo

```bash
git clone https://github.com/ggeorgol/ATACseq_course

cd ATACseq_course
```

## Create conda environment

There is a set of tools necessary to analyze high throughput sequencing data and ATAC-seq in particular. For this reason, we will create a virtual environment using the ANACONDA, conda for short, package manager. Specifically we are going to need the following tools:

* htslib 
* samtools
* bedTools
* bedops
* subread
* deepTools

```bash
module load gcc miniconda3
source $CONDA_PROFILE/conda.sh

conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge
conda config --set channel_priority strict

conda create -n atac python=3.10 htslib samtools deeptools bedtools bedops subread
```

## Download data

In this course we are going to work with ATAC-seq data generated by the [ENCODE project](https://www.encodeproject.org/matrix/?type=Experiment&control_type!=*&status=released&perturbed=false&assay_title=ATAC-seq). We will work with na√Øve and activated T-cells from a female adult with the following accession numbers: [ENCSR977LVI](https://www.encodeproject.org/experiments/ENCSR977LVI/), and [ENCSR558ZSN](https://www.encodeproject.org/experiments/ENCSR558ZSN/). We will use the alignment (BAM) files and the already generate peaks

```bash
DATADIR=${SCRATCH}/ATACseq_course/data # If you are on the cluster
mkdir -p ${DATADIR}$/{ENCSR977LVI,ENCSR558ZSN}
cd ${DATADIR}$/ENCSR558ZSN
wget https://www.encodeproject.org/files/ENCFF287DFF/@@download/ENCFF287DFF.bam
wget https://www.encodeproject.org/files/ENCFF218OSF/@@download/ENCFF218OSF.bam
wget https://www.encodeproject.org/files/ENCFF002MKC/@@download/ENCFF002MKC.bed.gz
wget https://www.encodeproject.org/files/ENCFF235RAD/@@download/ENCFF235RAD.bed.gz
```
